File: othersAlgorithm/binary_search_tree.py
class Node(object):
    def __init__(self, value):
        self.value = value
        self.parent = None
        self.left = None
        self.right = None

    def __str__(self):
        return f'<Node {self.value}>'

    __repr__ = __str__


class BinarySearchTree(object):
    def __init__(self):
        self.root = None
    
    def insert(self, value: int):
        def __insert(root: Node, value: int):
            if root is None:
                root = Node(value)
            else:
                if root.value > value:
                    __insert(root.left, value)
                elif root.value < value:
                    __insert(root.right, value)
                else:
                    raise ValueError(f'Value<{value}> exist in tree')
        __insert(self.root, value)

    def pre_travel(self, recursive: bool=False):
        """
        是否进行递归遍历
        """
        if recursive:
            def __pre_travel(root: Node):
                if root is None:
                    return
                print(root.value)
                if root.left is not None:
                    __pre_travel(root.left)
                if root.right is not None:
                    __pre_travel(root.right)
            __pre_travel(self.root)
        else:
           res = []
           p = self.root
           while p is not None or len(res)>0:
                while p is not None:
                    print(p.value)
                    res.append(p)
                    p = p.left
                if len(res)>0:
                    tmp_node = res[-1]
                    res.pop()
                    p = tmp_node.right

    def inorder_travel(self, recursive: bool=False):
        """
        中序遍历
        """
        if recursive:
            def __inorder_travel(root: Node):
                if root.left is not None:
                    __inorder_travel(root.left)
                print(root.value)
                if root.right is not None:
                    __inorder_travel(root.right)
            __inorder_travel(self.root)
        else:
            res = []
            p = self.root
            while p is not None or len(res) > 0:
                while p is not None:
                    res.append(p)
                    p = p.left
                if len(res)>0:
                    p = res[-1]
                    print(p.value)
                    res.pop()
                    p = p.right

    def postorder_travel(self):
        """
        后序遍历, 递归后序再补, 还没看
        """
        def __postorder_travel(root: Node):
            if root.left is not None:
                __postorder_travel(root.left)
            if root.right is not None:
                __postorder_travel(root.right)
            print(root.value)

    def floor_travel(self):
        """
        层遍历
        """
        queue = [self.root]
        while len(queue) > 0:
            _tmp_queue = []
            for item in queue:
                print(item, end=' ')
                if item.left is not None:
                    _tmp_queue.append(item.left)
                if item.right is not None:
                    _tmp_queue.append(item.right)
            print()
            queue = _tmp_queue

    def _find_parent(root: Node, value: int, predecessor: bool=True) -> Node:
        # 有父节点的时候, 直接使用递归即可
        if predecessor:
            parent = node.parent
            while parent is not None and parent.right != node:
                node = parent
                parent = node.parent
            return parent
        else:
            parent = node.parent
            while parent is not None and parent.left != node:
                node = parent
                parent = node.parent
            return parent

    def find_parent(self, predecessor: bool=True, value: int):
        """
        找前驱或者后继节点, 在节点里没有parent选项的时候
        """
        root = self.root
        p, first_parent, node = None, None, None
        if predecessor:
            while root:
                if root.value == value:
                    return node, p, first_parent
                p = root
                if root.value > value:
                    root = root.left
                else:
                    first_parent = root
                    root = root.right
            return None
        else:
            while root:
                if root.value == value:
                    return node, p, first_parent
                p = root
                if root.value < value:
                    root = root.right
                else:
                    first_parent = root
                    root = root.left
            return None

    def get_node(self, root: Node, is_left: bool=True):
        """
        找一个节点的左子树的右节点, 或者右子树扽左节点
        """
        if is_left:
            while node.right is not None:
                node = node.right
            return node
        else:
            while node.left is not None:
                node = node.left
            return node

    def predecessor(self, value: int):
        """
        找前驱节点
        """
        node, p, fp = self.find_parent(value=value)
        if node is None:
            return None
        if node.left is not None:
            return self.get_node(node, True)
        elif p is None of fp is None:
            return None
        elif node == p.right:
            return p
        return fp

    def successor(self, value: int):
        """
        找后继
        """
        node, p, fp = self.find_parent(False, value=value)
        if node is None:
            return None
        if node.right is not None:
            return self.get_node(node, False)
        elif p is None or fp is None:
            return None
        elif node == p.left:
            return p
        return fp

File: othersAlgorithm/go-examples/tree/binarySearchTree.go
package tree

import "fmt"

type Node struct {
	data  int
	left  *Node
	right *Node
}

func NewRoot(data int) *Node {
	return &Node{data: data}
}

func (t *Node) Insert(data int) *Node {
	if t == nil {
		t = NewRoot(data)
		return t
	}
	if data < t.data {
		t.left = t.left.Insert(data)
	} else {
		t.right = t.right.Insert(data)
	}
	return t
}

func (bt *Node) Pretravel(recursive bool) {
	if recursive {
		//递归来写, 前序遍历
		var inner func(node *Node)
		inner = func(node *Node) {
			if node == nil {
				return
			}
			fmt.Print(node.data, " ")
			if node.left != nil {
				inner(node.left)
			}
			if node.right != nil {
				inner(node.right)
			}
		}
		inner(bt)
	} else {
		// 用栈来做
		nodes := make([]*Node, 0)
		p := bt
		for p != nil || len(nodes) > 0 {
			for p != nil {
				fmt.Print(p.data, " \n")
				nodes = append(nodes, p)
				p = p.left
			}
			if len(nodes) > 0 {
				node := nodes[len(nodes)-1]
				nodes = nodes[:len(nodes)-1]
				p = node.right
			}
		}
	}
}

func (bt *Node) OrderTravel(recursive bool) {
	if recursive {
		var inner func(node *Node)
		inner = func(node *Node) {
			if node == nil {
				return
			}
			if node.left != nil {
				inner(node.left)
			}
			fmt.Print(node.data, " ")
			if node.right != nil {
				inner(node.right)
			}
		}
		inner(bt)
	} else {
		nodes := make([]*Node, 0)
		p := bt
		for p != nil || len(nodes) > 0 {
			for p != nil {
				nodes = append(nodes, p)
				p = p.left
			}
			if len(nodes) > 0 {
				node := nodes[len(nodes)-1]
				fmt.Println(node.data, " ")
				nodes = nodes[:len(nodes)-1]
				p = node.right
			}
		}
	}
}

func (bt *Node) PostTavel() {
	if bt == nil {
		return
	}
	if bt.left != nil {
		bt.left.PostTavel()
	}
	if bt.right != nil {
		bt.right.PostTavel()
	}
	fmt.Print(bt.data, " ")
}

func (bt *Node) FloorTravel() {
	queue := make([]*Node, 1)
	queue[0] = bt
	for len(queue) > 0 {
		tmp := make([]*Node, 0)
		for _, item := range queue {
			fmt.Print(item.data, " ")
			if item.left != nil {
				tmp = append(tmp, item.left)
			}
			if item.right != nil {
				tmp = append(tmp, item.right)
			}
		}
		queue = tmp
	}
}

// 有父亲节点
func (bt *Node) FindParent(node *Node, falg bool) {
	//中序遍历的前驱
	if falg {
		p := node.parent
		for p != nil && p.right != node {
			node = p
			p = node.parent
		}
		return p
	} else {
		p := node.parent
		for p != nil && p.left != node {
			node = p
			p = node.parent
		}
		return p
	}
}

// 终须
func (bt *Node) FindParentNoP(value int, flag bool) (node, firstParent, parent *Node) {
	root := bt
	if flag {
		for root != nil {
			if root.data == value {
				return node, p, firstParent
			}
			parent = root
			if p.data > value {
				root = root.left
			} else {
				firstParent = root
				root = root.right
			}
		}
	} else {
		for root != nil {
			if root.data == value {
				return
			}
			parent = root
			if root.data < value {
				root = root.right
			} else {
				firstParent = root
				root = root.left
			}
		}
	}
}

func getNode(node *Node, flag bool) {
	p := node
	if flag {
		for p.right != nil {
			p = p.right
		}
		return p
	} else {
		for p.left != nil {
			p = p.left
		}
		return p
	}
}

func (bt *Node) Predecessor(value int) *Node {
	node, firstParent, parent := bt.FindParentNoP(value, true)
	if node == nil {
		return nil
	}
	if node.left != nil {
		return getNode(node, true)
	} else if parent == nil || firstParent == nil {
		return nil
	} else if node == parent.right {
		return parent
	}
	return firstParent
}

func (bt *Node) Successos(value int) *Node {
	node, firstParent, parent := bt.FindParentNoP(value, false)
	if node == nil {
		return nil
	}
	if node.right != nil {
		return getNode(node, false)
	} else if parent == nil || firstParent == nil {
		return nil
	} else if node == parent.left {
		return parent
	}
	return firstParent
}

// 前序的遍历找前驱
func (bt *Node) PreP(node *Node) *Node {
	if node.parent != nil {
		return node.parent
	}
	if node == node.parent.right {
		if node.parent.left == nil {
			return node.parent
		} else {
			return getNode(node.parent.left, true)
		}
	}
}

func (bt *Node) SP(node *Node) *Node {
	if node.left != nil {
		return node.left
	} else if node.right != nil {
		return node.right
	} else {
		for (node.parent.right == nil || node.parent.right == node) && node != nil {
			node = node.parent
		}
		return node
	}
}

func (bt *Node) CommonFather(node1, node2 *Node) *Node {
	//寻找两个节点的公公祖先, 每个节点都做判断
	if hasNode(bt.left, node1) && hasNode(bt.left, node2) {
		return bt.left.CommonFather(node1, node2)
	}
	if hasNode(bt.right, node1) && hasNode(bt.right, node2) {
		return bt.right.CommonFather(node1, node2)
	}
	return bt
}

func hasNode(node1, node2 *Mode) bool {
	if node1 == nil {
		return false
	}
	if node1 == node2 {
		return true
	}
	return hasNode(node1.left, node2) || hasNode(node1.right, node2)
}

func (bt *Node) CommonFather2(node1, node2 *Node) *Node {
	if bt == nil {
		return nil
	}
	if bt.data == node1.data || bt.data == node2.data {
		return bt
	}
	l := bt.left.CommonFather2(node1, node2)
	r := bt.right.CommonFather2(node1, node2)
	if l && r {
		return bt
	}
	if l == nil {
		return r
	}
	return l
}
